File: taiwan_stock_app/frontend/lib/models/broker.dart
/// 券商帳戶連動資料模型

class BrokerAccount {
  final bool linked;
  final String? brokerType;
  final String? status;
  final String? accountNumber;
  final DateTime? lastSynced;

  BrokerAccount({
    required this.linked,
    this.brokerType,
    this.status,
    this.accountNumber,
    this.lastSynced,
  });

  factory BrokerAccount.fromJson(Map<String, dynamic> json) {
    return BrokerAccount(
      linked: json['linked'] ?? false,
      brokerType: json['broker_type'],
      status: json['status'],
      accountNumber: json['account_number'],
      lastSynced: json['last_synced'] != null
          ? DateTime.tryParse(json['last_synced'])
          : null,
    );
  }

  factory BrokerAccount.unlinked() => BrokerAccount(linked: false);
}

class BrokerPosition {
  final String symbol;
  final double quantity;
  final double avgCost;
  final double marketValue;
  final double unrealizedPnl;
  final DateTime? lastUpdated;

  BrokerPosition({
    required this.symbol,
    required this.quantity,
    required this.avgCost,
    required this.marketValue,
    required this.unrealizedPnl,
    this.lastUpdated,
  });

  factory BrokerPosition.fromJson(Map<String, dynamic> json) {
    return BrokerPosition(
      symbol: json['symbol'] ?? '',
      quantity: (json['quantity'] ?? 0).toDouble(),
      avgCost: (json['avg_cost'] ?? 0).toDouble(),
      marketValue: (json['market_value'] ?? 0).toDouble(),
      unrealizedPnl: (json['unrealized_pnl'] ?? 0).toDouble(),
      lastUpdated: json['last_updated'] != null
          ? DateTime.tryParse(json['last_updated'])
          : null,
    );
  }

  double get pnlPercent =>
      avgCost > 0 ? (unrealizedPnl / (avgCost * quantity)) * 100 : 0;
  bool get isProfit => unrealizedPnl > 0;
  bool get isLoss => unrealizedPnl < 0;
}

class BrokerLinkResponse {
  final int accountId;
  final String status;
  final String message;

  BrokerLinkResponse({
    required this.accountId,
    required this.status,
    required this.message,
  });

  factory BrokerLinkResponse.fromJson(Map<String, dynamic> json) {
    return BrokerLinkResponse(
      accountId: json['account_id'] ?? 0,
      status: json['status'] ?? '',
      message: json['message'] ?? '',
    );
  }

  bool get needsVerification => status == 'needs_2fa';
  bool get isActive => status == 'active';
  bool get isError => status == 'error';
}

File: taiwan_stock_app/backend/app/schemas/broker.py
"""
Broker Schemas — 券商連動 API 請求/回應模型
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ==================== 請求 ====================

class LinkAccountRequest(BaseModel):
    username: str = Field(..., min_length=1, max_length=100)
    password: str = Field(..., min_length=1, max_length=200)
    pin: str = Field(..., min_length=4, max_length=10)


class Verify2FARequest(BaseModel):
    account_id: int
    code: str = Field(..., min_length=4, max_length=10)


# ==================== 回應 ====================

class LinkAccountResponse(BaseModel):
    account_id: int
    status: str  # "active" | "needs_2fa" | "error"
    message: str


class BrokerStatusResponse(BaseModel):
    linked: bool
    broker_type: Optional[str] = None
    status: Optional[str] = None
    account_number: Optional[str] = None
    last_synced: Optional[datetime] = None


class BrokerPositionResponse(BaseModel):
    symbol: str
    quantity: float
    avg_cost: float
    market_value: float
    unrealized_pnl: float
    last_updated: Optional[datetime] = None

    model_config = {"from_attributes": True}


class BrokerPositionListResponse(BaseModel):
    positions: List[BrokerPositionResponse]
    total: int
